### PR TITLE
upload datapack framework

### PR DESCRIPTION
**Pull request by paologum**
based on a username that we hash, upload a datapack to that directory and create a json cache that the app can service when logging in new users.
Two main routes `/user-datapacks/:username/` and `/upload/:username`

First route grabs any datapacks that have been parsed in the users directory
Second route uploads a datapack, decrypts, parses and throws any error (with cleanup) with only one file.

TODO: sunset files and login functionality

### CODE REVIEW

File: server/src/routes.ts [hunk 1]
```diff
@@ -1,15 +1,177 @@
...
+  const datapackDir = path.join(userDir, "datapacks");
+  const decryptDir = path.join(userDir, "decrypted");
+  const filepath = path.join(datapackDir, filename);
+  const decryptedFilepathDir = path.join(decryptDir, filenameWithoutExtension);
+  const mapPackIndexFilepath = path.join(userDir, "MapPackIndex.json");
+  const datapackIndexFilepath = path.join(userDir, "DatapackIndex.json");
+  if (!/^(\.dpk|\.txt|\.map|\.mdpk)$/.test(ext)) {
+    reply.status(415).send({ error: "Invalid file type" });
+    return;
+  }
```
Comment (asivath) [inline server/src/routes.ts:81]: should also check MIME type since extensions can easily be faked

File: server/src/routes.ts [hunk 2]
```diff
@@ -1,15 +1,177 @@
...
+  if (!fs.existsSync(datapackDir)) {
+    fs.mkdirSync(datapackDir, { recursive: true });
+  }
+  const fileStream = file.file;
+  console.log("Uploading file: ", filename);
+  pump(fileStream, fs.createWriteStream(filepath));
+  try {
+    // must wait for the file to be written before decrypting
+    await new Promise<void>((resolve, reject) => {
+      pump(fileStream, fs.createWriteStream(filepath), (err) => {
```
Comment (asivath) [inline server/src/routes.ts:91]: two pumps?

File: server/src/routes.ts [hunk 3]
```diff
@@ -1,15 +1,177 @@
...
+  } catch (e) {
+    resetUploadDirectory(filepath, decryptedFilepathDir);
+    reply.status(500).send({ error: "Failed to save file with error: " + e });
+    return;
+  }
+  if (file.file.truncated) {
+    resetUploadDirectory(filepath, decryptedFilepathDir);
+    reply.status(413).send({ error: "File too large" });
+    return;
+  }
```
Comment (asivath) [inline server/src/routes.ts:108]: this should happen before uploading

File: server/src/routes.ts [hunk 4]
```diff
@@ -1,15 +1,177 @@
...
+  try {
+    const cmd =
+      `java -jar ${assetconfigs.decryptionJar} ` +
+      // Decrypting these datapacks:
+      `-d "${filepath}" ` +
+      // Tell it where to send the datapacks
+      `-dest ${decryptDir} `;
+    console.log("Calling Java decrypt.jar: ", cmd);
+    execSync(cmd, { stdio: "inherit" });
```
Comment (asivath) [inline server/src/routes.ts:117]: why not use async/await? Same with the other file operations

File: server/src/routes.ts [hunk 5]
```diff
@@ -1,15 +1,177 @@
...
+  if (fs.existsSync(path.join(userDir, "DatapackIndex.json"))) {
+    Object.assign(datapackIndex, JSON.parse(fs.readFileSync(path.join(userDir, "DatapackIndex.json")).toString()));
+  }
+  const mapPackIndex: MapPackIndex = {};
+  if (fs.existsSync(path.join(userDir, "MapPackIndex.json"))) {
+    Object.assign(mapPackIndex, JSON.parse(fs.readFileSync(path.join(userDir, "MapPackIndex.json")).toString()));
+  }
+  const indexResponse = { datapackIndex, mapPackIndex };
+  assertIndexResponse(indexResponse);
+  reply.status(200).send(indexResponse);
```
Comment (asivath) [inline server/src/routes.ts:50]: error handling for readFileSync or JSON.parse faling

File: server/src/routes.ts (hunk 3 above)
Comment (paologum) [reply server/src/routes.ts:108]: this is how `@fastify/multipart` is implemented, i tried looking into finding the size before but haven't seen anything that throws when the `fileSize` indicated in the options is exceeded until this bit of code catches it. what it does is read the file in to the limit and stops and sets truncate to true. we read the file through a buffer so i don't think that should cause much of a problem with overloading requests.

**Review by JacquiFreeElectron — COMMENTED** [review]
Looks good

File: server/src/routes.ts (hunk 1 above)
Comment (paologum) [reply server/src/routes.ts:93]: i can only really verify `application/octet-stream` and `text/plain` but i added them

**Review by asivath — APPROVED** [review]
looks good

**Pull request merged**